<center><h1>7SSG2059 Geocomputation 2017/18</h1></center>

<h1><center>Practical 6: Data Manipulation</h1></center>

<p><center><i>James Millington, 31 October 2017</i></center>



# Manipulating Data & Data Frames

As we've discussed in lectures, manipulating data can be a major component of data analysis. This week will look at some further ways to manipulate data that might be useful for you when analysing the data for your final report. 

Specifically we will:
1. recap on some useful sorting and selecting methods
2. see how we can combine two `DataFrames` together using common properties for further analysis
3. look at how we can group data for analysis (e.g. grouping LSOAs for borough-level analysis)

## Setup

As usual we will be using pandas, so we need to import the package:

In [1]:
import pandas as pd

We will start working with the initial LSOA data that we have been using previously:

In [2]:
my_df = pd.read_csv(
    'https://github.com/kingsgeocomp/geocomputation/blob/master/Data/LSOA%20Data.csv.gz?raw=true',
    compression='gzip', low_memory=False) # The 'low memory' option means pandas doesn't guess data types

Later we'll look at how we can add more data to this `DataFrame` later, but first let's just check what columns of data we have:

In [3]:
my_df.columns

Index([u'LSOA11CD', u'LSOA11NM', u'MSOA11CD', u'MSOA11NM', u'LAD11CD',
       u'LAD11NM', u'RGN11CD', u'RGN11NM', u'USUALRES', u'HHOLDRES',
       u'COMESTRES', u'POPDEN', u'HHOLDS', u'AVHHOLDSZ', u'geometry',
       u'GreenspaceArea', u'RoadsArea', u'Owned', u'SharedOwnshp',
       u'SocialRented', u'PrivateRented', u'RentFree', u'ORroomsM1',
       u'ORbedsM1', u'RoomsHH', u'BedsHH', u'White', u'MultiEthnic', u'Asian',
       u'Black', u'OtherEthnic', u'MedianIncome', u'SumPrice', u'MeanPrice',
       u'MedianPrice', u'EntireHome', u'PrivateRoom', u'SharedRoom',
       u'SmallHost', u'MultiHost', u'PropertyCount', u'Area'],
      dtype='object')

Okay, now we have our data loaded and we've reminded ourselves of what the data set contains (maybe by consulting the [metadata](https://github.com/kingsgeocomp/geocomputation/raw/master/Data/LSOA_metadata.xlsx)) we can move on. 

# Recap: Sorting and Selecting 

## Finding Rows in the Data

You should remember from Week 3 that we can find out _where_ (i.e. which LSOA) the maximum value in the data occurs using code like this:
```python
my_df[my_df.POPDEN == my_df.POPDEN.max()]
```

### Task:
Write some code to list all of the LSOAs where the population density is more than two standard deviations greater than the mean population density of all London LSOAs: 

In [4]:
my_df[my_df.POPDEN > (my_df.POPDEN.mean() + (2 * my_df.POPDEN.std()))]

LSOA11CD                     LSOA11NM   MSOA11CD  \
2     E01000003          City of London 001C  E02000001   
44    E01000048    Barking and Dagenham 021B  E02000022   
47    E01000051    Barking and Dagenham 021D  E02000022   
533   E01000544                   Brent 034A  E02000126   
536   E01000547                   Brent 028C  E02000120   
623   E01000637                   Brent 022B  E02000114   
835   E01000853                  Camden 025A  E02000190   
887   E01000905                  Camden 012D  E02000177   
899   E01000917                  Camden 027C  E02000192   
914   E01000932                  Camden 013D  E02000178   
917   E01000935                  Camden 020E  E02000185   
922   E01000940                  Camden 024D  E02000189   
923   E01000941                  Camden 025C  E02000190   
924   E01000942                  Camden 025D  E02000190   
926   E01000944                  Camden 023A  E02000188   
931   E01000949                  Camden 023C  E02000188   
933   E01000951                  Camden 023E  E02000188   
936   E01000954                  Camden 022C  E02000187   
937   E01000955                  Camden 022D  E02000187   
941   E01000959                  Camden 019E  E02000184   
1336  E01001356                  Ealing 033C  E02000270   
1666  E01001703               Greenwich 011E  E02000323   
1687  E01001726                 Hackney 004B  E02000348   
1689  E01001728                 Hackney 029B  E02006921   
1693  E01001732                 Hackney 019C  E02000363   
1706  E01001746                 Hackney 016B  E02000360   
1717  E01001757                 Hackney 016C  E02000360   
1740  E01001783                 Hackney 026D  E02000370   
1741  E01001784                 Hackney 027E  E02000371   
1747  E01001790                 Hackney 013D  E02000357   
...         ...                          ...        ...   
4643  E01032513             Westminster 010G  E02000969   
4666  E01032637               Southwark 003F  E02000809   
4667  E01032638               Southwark 003G  E02000809   
4669  E01032640               Southwark 003H  E02000809   
4671  E01032719               Southwark 014G  E02000820   
4678  E01032743                  Ealing 010F  E02000247   
4686  E01032770           Tower Hamlets 032B  E02006853   
4689  E01032773           Tower Hamlets 032D  E02006853   
4691  E01032775           Tower Hamlets 031F  E02000894   
4695  E01032779           Tower Hamlets 028H  E02000891   
4697  E01032781           Tower Hamlets 033C  E02006854   
4700  E01032784           Tower Hamlets 004G  E02000867   
4702  E01032786           Tower Hamlets 010F  E02000873   
4703  E01032787           Tower Hamlets 011F  E02000874   
4705  E01032789  Hammersmith and Fulham 023E  E02000394   
4707  E01032803               Southwark 003J  E02000809   
4709  E01032834               Southwark 003K  E02000809   
4711  E01033030                Hounslow 025G  E02000550   
4732  E01033208                 Lambeth 004G  E02000621   
4747  E01033487               Islington 011F  E02000564   
4753  E01033493               Islington 006F  E02000559   
4785  E01033598             Westminster 009F  E02000968   
4786  E01033599             Westminster 023G  E02000982   
4788  E01033601             Westminster 009G  E02000968   
4789  E01033602             Westminster 009H  E02000968   
4790  E01033603             Westminster 009I  E02000968   
4792  E01033605             Westminster 009K  E02000968   
4793  E01033606             Westminster 014F  E02000973   
4795  E01033608             Westminster 007G  E02000966   
4808  E01033710                 Hackney 028E  E02006918   

                        MSOA11NM    LAD11CD                 LAD11NM  \
2             City of London 001  E09000001          City of London   
44      Barking and Dagenham 021  E09000002    Barking and Dagenham   
47      Barking and Dagenham 021  E09000002    Barking and Dagenham   
533                    Brent 034  E09000005                   Br

Hopefully, your code returns you 198 LSOAs.

Good, now what about if we want to find the top 10 LSOAs in terms of population density, and examine how many households there are in those LSOAs? First we'll do this with a couple of lines of code, then we'll see how we can _chain_ together a series of methods. 

First the code that uses several lines of code:

In [5]:
sort_df = my_df.sort_values(by='POPDEN', ascending=False)
sort_df = sort_df.head(10)
sort_df = sort_df[['LSOA11NM','POPDEN','HHOLDS']]

print(sort_df)

                         LSOA11NM  POPDEN  HHOLDS
2777  Kensington and Chelsea 021C   684.7     514
4747               Islington 011F   502.5     595
4618             Westminster 024E   463.2     950
4689           Tower Hamlets 032D   461.8     852
4631             Westminster 022D   437.0     905
4194           Tower Hamlets 025C   399.0     637
4793             Westminster 014F   394.1     614
4538             Westminster 017A   391.5     958
4705  Hammersmith and Fulham 023E   384.9     586
1861  Hammersmith and Fulham 015B   378.7     875


Let's clarify exactly what we did there: :
1. Sorted `my_df` in descending order on `POPDEN`, assigning the output to `sort_df`;
2. Take the first ten values of `sort_df` and assign to `sort_df`;
3. Select only the three columns we want and assign to `sort_df`;
4. Print out the columns specified by the list.

Now let's do the same thing but with lines 1-3 combined into a single line:

In [6]:
sort_df = my_df.sort_values(by='POPDEN', ascending=False).head(10)[['LSOA11NM','POPDEN','HHOLDS']]

print(sort_df)

                         LSOA11NM  POPDEN  HHOLDS
2777  Kensington and Chelsea 021C   684.7     514
4747               Islington 011F   502.5     595
4618             Westminster 024E   463.2     950
4689           Tower Hamlets 032D   461.8     852
4631             Westminster 022D   437.0     905
4194           Tower Hamlets 025C   399.0     637
4793             Westminster 014F   394.1     614
4538             Westminster 017A   391.5     958
4705  Hammersmith and Fulham 023E   384.9     586
1861  Hammersmith and Fulham 015B   378.7     875


Take a look at that line of code and check you can see how the different lines previously have been combined. 

Let's pull it apart step-by-step at the code level:

* The first step in this process is `my_df.sort_values` -- you can probably guess what this does: it sorts the data frame!
* The parameters passed to the `sort_values` function are `by`, which is the column on which to sort, and `ascending=False`, which gives us the data frame sorted in _descending_ order!
* The output of `my_df.sort(...)` is a _new_ data frame, which means that we can simply add `.head(10)` to get the first ten rows of the newly-sorted data frame.
* And the output of `my_df.sort(...).head(...)` is yet _another_ data frame, which means that we can print out the values of selected columns using the 'dictionary-like' syntax: we use the outer set of square brackets (`[...]`) to tell pandas that we want to access a subset of the top-10 data frame, and we use the inner set of square brackets (`['LSOA11NM','POPDEN','HHOLDS']`) to tell pandas which columns we want to see.

I'd say 'simples, right?' but that's obviously _not_ simple. It _is_, however, very, very _elegant_ because it's quite clear (once you get past the way that lots of methods can be chained together) and it's very succinct (we did all of that in _one_ line of code!).

### Task:
In a single line of code create a new `df` containing information about the name, population density and number of usual residents for the seven least populated LSOAs (in terms of usual residents). Then use another line of code to print the new `df`: 

In [7]:
sort_df2 = my_df.sort_values(by='USUALRES', ascending=True).head(7)[['LSOA11NM','POPDEN','USUALRES']]
print(sort_df2)

                 LSOA11NM  POPDEN  USUALRES
3     City of London 001E    52.0       985
4784     Westminster 023F   168.9       990
4719      Wandsworth 002F    30.8      1012
4819       Greenwich 004E    49.7      1017
4762          Newham 034G    66.3      1021
4720      Wandsworth 004E    99.2      1028
4746       Islington 011E   155.3      1036


## Taking a Random Sample of Data

Of course, sometimes you don't want a particular range of data, you want a _random sample_ so that you can either 
a. get a better sense of the data, or 
b. perform some kind of test with a subsample before replicating on the full data set. 

Pandas has [got you covered](http://pandas.pydata.org/pandas-docs/version/0.18.1/indexing.html#selecting-random-samples) with a huge range of options, including sampling with replacement, sample weights, row numbers and a fraction of the data set. 

Let's look at some simple examples:

In [9]:
my_df.sample(n=5)[ ['LSOA11NM','POPDEN','USUALRES'] ] # Sample of size 5

LSOA11NM  POPDEN  USUALRES
2658      Islington 016C   230.6      2334
3130       Lewisham 009A    86.3      1461
4212  Tower Hamlets 021C   174.7      1822
894          Camden 001D    17.7      1589
2370     Hillingdon 018A    74.2      1436

In [10]:
my_df.sample(n=5)[ ['LSOA11NM','POPDEN','USUALRES'] ] # This will not give you the same sample

LSOA11NM  POPDEN  USUALRES
1213     Ealing 031C    71.3      1538
2140     Harrow 009D    12.3      1672
3957  Southwark 001E    65.6      1291
950      Camden 016E   184.7      1505
973     Croydon 014A    39.9      1978

Note that even though the two lines of code above are identical we return a different (random) sample of rows. This is useful but what if we want to give our code to someone else to that they would get the same (random) sample of rows?  To do this we can specify the `random_state` argument:

In [13]:
my_df.sample(n=5, random_state=2)[ ['LSOA11NM','POPDEN','HHOLDS'] ] 

LSOA11NM  POPDEN  HHOLDS
549                  Brent 023D    77.9     727
2854  Kingston upon Thames 017B    22.2     669
1978              Haringey 010A   139.3     552
4460            Wandsworth 016A    21.1     706
683                Bromley 010B    10.5     439

By specifying the same value for `random_state` we will get the same sample: 

In [14]:
my_df.sample(n=5, random_state=2)[ ['LSOA11NM','POPDEN','HHOLDS'] ] 

LSOA11NM  POPDEN  HHOLDS
549                  Brent 023D    77.9     727
2854  Kingston upon Thames 017B    22.2     669
1978              Haringey 010A   139.3     552
4460            Wandsworth 016A    21.1     706
683                Bromley 010B    10.5     439

And using a different value for `random_state` gives us a different sample:

In [15]:
my_df.sample(n=5, random_state=3)[ ['LSOA11NM','POPDEN','HHOLDS'] ] 

LSOA11NM  POPDEN  HHOLDS
994                Croydon 020B   101.5     899
261                 Barnet 011B    23.8     511
20    Barking and Dagenham 008B   102.4     474
3051               Lambeth 032C   110.4     813
3217              Lewisham 012A   106.2     794

We can also specify the fraction of the `DataFrame` we want to sample, rather than an absolute number of observations (think about why this is useful for when we don't know what size `DataFrame` our code might be used with):

In [16]:
my_df.sample(frac=0.002)[ ['LSOA11NM','POPDEN','USUALRES'] ] # Sample a fraction of the rows (here 0.2%)

LSOA11NM  POPDEN  USUALRES
1481         Enfield 012B    85.3      1549
2527        Hounslow 013B    51.5      1691
4308  Waltham Forest 005C    60.2      1663
1112         Croydon 018C    95.4      1563
3148        Lewisham 022A   103.9      1670
4217   Tower Hamlets 017D   172.5      1444
1779         Hackney 004E   154.3      1740
3457          Newham 010E   174.6      2395
1362          Ealing 028D   154.8      1791
1617       Greenwich 019B    83.1      1733

Finally, the code above has automatically been sampling rows of data, but we can also sample columns by specifying the `axis` of the `DataFrame` we want to sample: 

In [17]:
my_df.sample(n=5, axis=1).head(10)

PrivateRoom                   LSOA11NM  RentFree  Black  PropertyCount
0          1.0        City of London 001A      35.0   11.0           17.0
1          5.0        City of London 001B      28.0    4.0           18.0
2          5.0        City of London 001C      17.0   45.0           12.0
3         14.0        City of London 001E       8.0  100.0           51.0
4          1.0  Barking and Dagenham 016A       2.0  177.0            1.0
5          6.0  Barking and Dagenham 015A       6.0  389.0            7.0
6          3.0  Barking and Dagenham 015B       6.0  574.0            4.0
7          1.0  Barking and Dagenham 016B       4.0  228.0            2.0
8          7.0  Barking and Dagenham 015C       5.0  580.0           18.0
9          3.0  Barking and Dagenham 016C      12.0  143.0            3.0

# Combining Data

Up until this point we have been working with a dataset of ~48 variables (columns) for the LSOAs. But what if we have additional data for LSOAs that we want to work with together with our original data, for example to look for correlations between variables. Here we will look at how to combine two datasets that have data for individual LSOAs:
1. our original data
2. data for air quality in each LSOA

Combining these data would be useful, for example, to examine relationships between air quality and socio-economic and other variables. 

In this practical we will look at how to `merge` Pandas dataframes. There's also `join` and `concatenate` function. Each of these functions are slightly different:
- `merge` enables us to [combine](http://pandas.pydata.org/pandas-docs/stable/merging.html#database-style-dataframe-joining-merging) two dataframes based on a column that is common between them
- `join` is used to [combine](http://pandas.pydata.org/pandas-docs/stable/merging.html#joining-on-index) two dataframes when they share a common `index` (e.g. a `DateTimeIndex` in timeseries data)
- `concatenate` [combines](http://pandas.pydata.org/pandas-docs/stable/merging.html#concatenating-objects) dataframes regardless of common attributes. 

We'll then look at how to analyse variables in the combined `DataFrame` we produce using correlation later in the module. 

## Air Quality Data

Metadata about the air quality data are included in the [metadata](https://github.com/kingsgeocomp/geocomputation/raw/master/Data/LSOA_metadata.xlsx) file. The data themselves are hosted online and can be read using:

In [18]:
aq_df = pd.read_csv(
    'https://github.com/kingsgeocomp/geocomputation/blob/master/Data/LSOA_AirQuality.csv.gz?raw=true',
    compression='gzip', low_memory=False) # The 'low memory' option means pandas doesn't guess data types

In [19]:
aq_df.head()

LSOA11CD  PM25mean  PM25min  PM25max  PM25sd  PM10mean  PM10min  PM10max  \
0  E01000001   18.6339  17.6449  23.1297  1.0451   29.6152  27.2757  40.9147   
1  E01000002   17.8920  17.5738  20.0114  0.4147   27.8759  27.1356  32.2687   
2  E01000003   17.8823  17.5894  20.2573  0.3984   27.8970  27.1989  33.6703   
3  E01000005   18.4486  17.5878  30.7432  1.4506   29.7238  27.2533  59.8320   
4  E01000006   15.7733  15.6290  15.9539  0.0666   24.5767  24.3273  25.0226   

   PM10sd   NOxmean   NOxmin    NOxmax    NOxsd  NO2mean   NO2min    NO2max  \
0  2.5343  116.7335  74.3843  368.4532  46.8249  57.8972  43.0877  127.2501   
1  0.9467   86.3397  72.6057  191.7217  17.0302  47.6356  42.3820   80.7450   
2  1.0226   86.1607  74.2665  189.0205  17.5927  47.3078  42.9148   81.4941   
3  3.8027  132.1101  76.5674  893.0905  89.5282  61.1162  43.7231  234.1228   
4  0.1474   51.3147  46.9127   57.4380   2.1936  32.5846  31.1267   34.9157   

     NO2sd  
0  14.9486  
1   6.3878  
2   6.0950  
3  23.2659  
4   0.7742

### Task

Familiarise yourself with the data you have just loaded in and compare it to the data we have worked with previously. To do this you might:
1. check the column names and data types of the air quality data file and compare to the metadate file (hint: use `info()` method for `DataFrame`s - you might need to google this)  
2. calculate descriptive statistics for the air quality data 
3. compare the column names of the air quality dataset with the original LSOA dataset
4. compare the shapes of the two `DataFrames`

In [20]:
#1 code here
aq_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4835 entries, 0 to 4834
Data columns (total 17 columns):
LSOA11CD    4835 non-null object
PM25mean    4835 non-null float64
PM25min     4835 non-null float64
PM25max     4835 non-null float64
PM25sd      4835 non-null float64
PM10mean    4835 non-null float64
PM10min     4835 non-null float64
PM10max     4835 non-null float64
PM10sd      4835 non-null float64
NOxmean     4835 non-null float64
NOxmin      4835 non-null float64
NOxmax      4835 non-null float64
NOxsd       4835 non-null float64
NO2mean     4835 non-null float64
NO2min      4835 non-null float64
NO2max      4835 non-null float64
NO2sd       4835 non-null float64
dtypes: float64(16), object(1)
memory usage: 642.2+ KB


In [21]:
#2 code here
aq_df.describe()

PM25mean      PM25min      PM25max       PM25sd     PM10mean  \
count  4835.000000  4835.000000  4835.000000  4835.000000  4835.000000   
mean     15.933692    15.663149    18.791063     0.417154    25.081174   
std       0.747620     0.644695     2.767815     0.316205     1.332229   
min      14.320600    14.035100    14.626100     0.028200    22.208700   
25%      15.399250    15.204750    17.202900     0.202350    24.115550   
50%      15.819100    15.575500    18.291300     0.349300    24.886200   
75%      16.373850    16.054050    19.731650     0.537950    25.853150   
max      19.523700    18.641100   110.827100     2.279000    31.353200   

           PM10min      PM10max       PM10sd      NOxmean       NOxmin  \
count  4835.000000  4835.000000  4835.000000  4835.000000  4835.000000   
mean     24.427725    32.410070     1.074251    60.067193    49.747635   
std       1.034080     6.057279     0.808448    16.524965    10.552390   
min      21.632900    22.785200     0.063400    27.853100    23.437500   
25%      23.677600    28.535350     0.513050    48.577450    42.097450   
50%      24.314600    31.260400     0.906100    56.812700    48.466900   
75%      25.096200    34.886000     1.413200    68.332950    56.316650   
max      28.658600   119.399400     5.808300   168.368200    93.074000   

            NOxmax        NOxsd      NO2mean       NO2min       NO2max  \
count  4835.000000  4835.000000  4835.000000  4835.000000  4835.000000   
mean    175.898430    16.469051    36.015590    32.388520    71.426087   
std     163.822116    13.330336     6.516267     4.674863    28.459356   
min      35.656800     0.975100    21.684400    19.153200    24.710500   
25%     107.646950     7.188150    31.362050    29.002650    52.610150   
50%     150.183900    13.043000    35.043200    31.980100    66.588400   
75%     217.482500    21.994050    39.678750    35.510300    84.813500   
max    9237.618200   137.826800    70.707000    49.975300   643.045700   

             NO2sd  
count  4835.000000  
mean      5.442446  
std       3.934605  
min       0.385900  
25%       2.606750  
50%       4.584300  
75%       7.271400  
max      29.369500

In [22]:
#3 code here
print(aq_df.columns.values)
print(my_df.columns.values)

['LSOA11CD' 'PM25mean' 'PM25min' 'PM25max' 'PM25sd' 'PM10mean' 'PM10min'
 'PM10max' 'PM10sd' 'NOxmean' 'NOxmin' 'NOxmax' 'NOxsd' 'NO2mean' 'NO2min'
 'NO2max' 'NO2sd']
['LSOA11CD' 'LSOA11NM' 'MSOA11CD' 'MSOA11NM' 'LAD11CD' 'LAD11NM' 'RGN11CD'
 'RGN11NM' 'USUALRES' 'HHOLDRES' 'COMESTRES' 'POPDEN' 'HHOLDS' 'AVHHOLDSZ'
 'geometry' 'GreenspaceArea' 'RoadsArea' 'Owned' 'SharedOwnshp'
 'SocialRented' 'PrivateRented' 'RentFree' 'ORroomsM1' 'ORbedsM1' 'RoomsHH'
 'BedsHH' 'White' 'MultiEthnic' 'Asian' 'Black' 'OtherEthnic'
 'MedianIncome' 'SumPrice' 'MeanPrice' 'MedianPrice' 'EntireHome'
 'PrivateRoom' 'SharedRoom' 'SmallHost' 'MultiHost' 'PropertyCount' 'Area']


In [23]:
#4 code here
print("aq_df has {0} rows and {1} columns".format(aq_df.shape[0],aq_df.shape[1]))
print("my_df has {0} rows and {1} columns".format(my_df.shape[0],my_df.shape[1]))

aq_df has 4835 rows and 17 columns
my_df has 4835 rows and 42 columns


From your exploration of the new data and comparison with the original LSOA data you might notice a few things:
1. They have the same number of rows
2. They have different numbers of columns
3. They share one column name in common (`LSOA11CD`)

Check you can see these observations for yourself. 

## Merge

If we have a column in each of two `DataFrames` that contains the same identifier (column) for the other variables (columns) in the data, we can use the common identifier to define how the two `DataFrames` are joined together. For example, the data we are working with are for LSOAs (distinct geographical regions) - if any additional data we have is also for LSOAs, as long as we we have a common way of identifying the LSOAs in both `DataFrames` we can `merge` the `Data Frames`. 

Hopefully from the task above that we have a common identifier in both the orginal data `my_df` and the new air qualiy data `aq_df`: `LSOA11CD`. The `LSOA11CD` is a unique identifier code for each LSOA. We can use this to match rows of data in `my_df` (each of which is for a particular LSOA) with the corresponding rows in `aq_df` (which are also for individual LSOAs).

With the common identifier identified, we now need to decide what type of join we want to do. Recall from this week's lecture that there are four main types of 'join':
1. left
2. right
3. outer
4. inner

We could use any of the above depending on our objectives. 

Here we'll do a **left join**, where the left `df` will be the original data and the right `df` will be our new air quality data. This seems appropriate so that we don't modify the original data too much (therby potentially messing up some of our code):

In [24]:
#merge the two data frames 
merge_df = pd.merge(my_df, aq_df, how = 'left', on = 'LSOA11CD')

Okay, now let's check what the columns are in the new `DataFrame` we just created: 

In [25]:
print(merge_df.columns)

Index([u'LSOA11CD', u'LSOA11NM', u'MSOA11CD', u'MSOA11NM', u'LAD11CD',
       u'LAD11NM', u'RGN11CD', u'RGN11NM', u'USUALRES', u'HHOLDRES',
       u'COMESTRES', u'POPDEN', u'HHOLDS', u'AVHHOLDSZ', u'geometry',
       u'GreenspaceArea', u'RoadsArea', u'Owned', u'SharedOwnshp',
       u'SocialRented', u'PrivateRented', u'RentFree', u'ORroomsM1',
       u'ORbedsM1', u'RoomsHH', u'BedsHH', u'White', u'MultiEthnic', u'Asian',
       u'Black', u'OtherEthnic', u'MedianIncome', u'SumPrice', u'MeanPrice',
       u'MedianPrice', u'EntireHome', u'PrivateRoom', u'SharedRoom',
       u'SmallHost', u'MultiHost', u'PropertyCount', u'Area', u'PM25mean',
       u'PM25min', u'PM25max', u'PM25sd', u'PM10mean', u'PM10min', u'PM10max',
       u'PM10sd', u'NOxmean', u'NOxmin', u'NOxmax', u'NOxsd', u'NO2mean',
       u'NO2min', u'NO2max', u'NO2sd'],
      dtype='object')


And let's check from a sample of the data how the rows look: 

In [26]:
merge_df.sample(n=5, random_state=3)[ ['LSOA11CD','POPDEN','HHOLDS', 'PM25mean', 'PM25min', 'PM25max', 'PM25sd'] ] 

LSOA11CD  POPDEN  HHOLDS  PM25mean  PM25min  PM25max  PM25sd
994   E01001013   101.5     899   15.9463  15.7213  19.8171  0.5193
261   E01000266    23.8     511   15.3256  15.1957  17.6231  0.2304
20    E01000022   102.4     474   15.3746  15.3045  15.4978  0.0383
3051  E01003119   110.4     813   16.3320  15.8218  19.7380  0.6776
3217  E01003288   106.2     794   16.1950  15.9349  18.4951  0.4249

Hopefully it looks good so far. Let's review what we did the with `pd.merge` function above:
```python
pd.merge(my_df, aq_df, how = 'left', on = 'LSOA11CD')
```

1. `my_df` is the left `df`
2. `aq_df` is the right `df` 
3. `how` defines how the join is to be done
4. `on` is the column we want to use as the common identifier to 'join on' 

So above each value in the `LSOA11CD` column in `aq_df` is matched with the same value in the `LSOA11CD` column in `my_df` and the rows those values are found in are combined. The figure below illustrates the process. 

![Illustration of the Pandas merge function](http://pandas.pydata.org/pandas-docs/stable/_images/merging_merge_on_key_left.png)

Check you understand how something similar to the image above has been done for our LSOA data. Remember you can read [the documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html#database-style-dataframe-joining-merging) for more detailed explanation. 

Even though we used the pandas `merge` function here, we are doing what we called a _join_ in the lecture; the only difference between pandas `merge` and `join` is that the former uses a common column whereas the latter uses a common index. The 'merge' column can do all four of the joins we have considered (left, right, outer and inner).   

To check if there were any missing values introduced into our new `DataFrame` we can do [a quick check](https://stackoverflow.com/a/29530601):

In [28]:
merge_df.isnull().values.any()

False

Hopefully you received a `False` response! If so, this is more evidence the join worked (if not you might want to check what you did above and ask for help). 

Let's save these data for later - they may be useful for your final report!

In [30]:
merge_df.to_csv("LondonLSOAData.csv", index=False)
#or
#merge_df.to_csv("LondonLSOAData.csv.gz", compression='gzip', index=False)

(If you really want to check what the join has done, you might open the file you just saved in Excel to have a look)

### Task:

Explore the air quality data to get an understanding of what they might show you in relation to other variables in the data set. For example:

1. Find the population densities of the LSOA with highest mean values for each of the four pollutants.  

2. Create a single barplot to compare the distributions of the mean values of each of the four pollutants  

3. Create four scatter plots within a single figure (use a loop) to visualise the relationship between area within 250m of a major road and the maximum values of each pollutant   

In [ ]:
#code for task 1 here

In [ ]:
#code for task 2 here

In [ ]:
#code for task 3 here

## Grouping Data

4.2.4	Now that we have our Borough column we can tell pandas to group the data using the values in that column (alternatively we could do some filtering/selecting on the column as we’ll see in section 4.3). The pandas library has another data structure known as DataFrameGroupBy which is useful in this situation (read more here). We can create one of these data structures for to group our boroughs using the groupby method:

### Matching on the Start of a Word

This is easier in the short-term because we can just say "find the Hackney GeoLabels" but it's less flexible in the long-term because we can't actually use the GeoLabel column as a way to group our results (because each GeoLabel is _still_ unique, so grouping will group by LSOA).

But let's take a look at how that works anyway:

In [ ]:
df.ix[df.GeoLabel.str.startswith('Hackney'),['CDU','GeoCode','GeoLabel','Total']].head(10)

You'll see that that gave us the Hackney LSOAs, but without modifying any of the data. This is also the first time you've seen the `<data frame>.ix` notation which allows us to combine row and and column access via a mix of integer and label. In other words, you can use this to [select _any_ kind of subset you like](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.ix.html).

Just to make it really obvious what I did:
```python
<data frame>.ix[ <row selection criteria>, <column selection criteria> ].head()
```
So this is:
1. Select rows where `<data frame>.<series name>.str.startswith(<search string>)` (treats the LSOA GeoLable as a string and the searches for strings that start with...)
2. Select columns in the list `['CDU','GeoCode','GeoLabel','Total']`
3. Return the first 10 rows using `head()`

You could _also_ use integer selection on the columns as you would with any normal list: e.g. `xrange(2,5)`.

### Selecting by Match

Now we can take a 'cut' of the data by selecting only some London-based boroughs for further analysis... we’re going to arbitrarily select K&C, Hackney and Barking because I know they’re quite different boroughs, but you could also use the data to make this selection, right? For instance, I could ask pandas to help me pick the three boroughs that are the furthest apart in terms of their Group 1 means...

Anyway, here’s the code to select a cut:
```python
# sdf = subsetted data frame

# Select where the LA value 'is in' one of our pre-defined list
sdf = df.loc[df.LA.isin(['Kensington and Chelsea','Hackney','Barking and Dagenham'])]

# Remove the remaining unused categories
sdf.LA = sdf.LA.cat.remove_unused_categories()

# And a simple check to see how many categorical values are left
print("sdf now contains: {0} values".format(sdf.LA.describe().unique()[1]))
```
You’ll notice two unusual bits of code in there that need some explanation:
* To 'select  multiple' we need to write: `<data frame>.<series name>.isin([...])`, so we can't just write `<data frame>.<series name>==[...]` unfortunately.
* We also have this line: `<data frame>.<column name> = <data frame>.<column name>.cat.remove_unused_categories()`. This should be fairly self-explanatory, but it’s because by default pandas doesn’t update the list of valid categories (i.e. Local Authorities) just because we filtered out boroughs that weren’t of interest. We therefore need to update the Series so that Seaborn doesn’t include a bunch of empty categories in when we make our plots.

In [6]:
my_df.LAD11NM.unique()

array(['City of London', 'Barking and Dagenham', 'Barnet', 'Bexley',
       'Brent', 'Bromley', 'Camden', 'Croydon', 'Ealing', 'Enfield',
       'Greenwich', 'Hackney', 'Hammersmith and Fulham', 'Haringey',
       'Harrow', 'Havering', 'Hillingdon', 'Hounslow', 'Islington',
       'Kensington and Chelsea', 'Kingston upon Thames', 'Lambeth',
       'Lewisham', 'Merton', 'Newham', 'Redbridge', 'Richmond upon Thames',
       'Southwark', 'Sutton', 'Tower Hamlets', 'Waltham Forest',
       'Wandsworth', 'Westminster'], dtype=object)

In [ ]:
boroughs = df.groupby('Borough')

4.2.5	The boroughs `DataFrameGroupBy` object is a special type of DataFrame that has additional methods available based on what groups we have specified (in this case the borough). For example, when we use the head method it looks similar to a normal DataFrame

In [ ]:
boroughs.head()

But when we try to get the shape of the object we find it's slightly different from a normal DataFrame (you should get an error): 

In [ ]:
boroughs.shape[0]

4.2.6	So we can’t use shape to find out how many elements in the boroughs DataFrameGroupBy object, but we can use our old favourite function len (which works pretty much anywhere!). Compare the output for the next two lines of code:

In [ ]:
len(boroughs) 

In [ ]:
len(df)

The length of boroughs is the number of groups in the DataFrameGroupBy object, whereas the length of df is the number of rows in the DataFrame object. Check you understand the difference! We can tell from this that there are 33 boroughs (groups) and 4835 LSOAs (rows).
4.2.7	The difference between df and boroughs also results in different output for other methods. Compare the output of the following:

In [ ]:
boroughs["GEO_LABEL"].count()

#or

df.boroughs.count()

See how the count method for the DataFrameGroupBy object gives the count of LSOAs within in each borough (group) whereas the count method for the DataFrame object simply gives the count of the total number of LSOAs (rows). Note, the following two lines of input code do exactly the same as the last two but with slightly different notation (known as dot notation): 

In [ ]:
boroughs.GEO_LABEL.count()
df.GEO_LABEL.count()

Dot notation looks a bit nicer but isn’t always as flexible as using [].
1.1.1	There are other methods we can use on DataFrameGroupBy, for example get_group gets the data (for LSOAs) for just one of the groups (boroughs):

In [ ]:
boroughs.get_group("City of London")  

1.1.2	Using the DataFrameGroupBy object also allows us to describe the data by group (rather than for all of the LSOAs as we did before). For example, to find the mean values for the columns by borough we can use the aggregate method. 

In [ ]:
bMeans = boroughs.aggregate(np.mean) 
print(bMeans)

Note how the aggregate method makes a call to the numpy function mean (the code at In [72] assumes you did import numpy as np). The aggregate method returns a DataFrame; check this by using type(bMeans). Check you understand what has been produced here; the mean of all columns in our original NS-SeC dataset but aggregated (grouped) by borough. (If we had added our Boroughs column to Excel we would use a Pivot Table to get similar output). 

1.1.3	Let’s see how this all might be useful for answering a geographical question. Say we want to calculate what proportion of the population of the Borough of Harrow that is ‘lower managerial’ (Group 2). The code to do this is shown in below

In [ ]:
#What proportion of population of Harrow is lower managerial?  (code in Figure 2 in practical handout)
boroughs = df.groupby('Borough')                                     #as In[62]  
bSums = boroughs.aggregate(np.sum)                                   #sum of columns grouped by borough
harrow_sumG2 = bSums.Group2.loc["Harrow"]                            #sum of Group 2 for LSOAs in Harrow  #equivalent using dot notation is harrow_sumG2 = harrow.Group2.sum()   
harrow_sumTot = bSums.Total.loc["Harrow"]                            #sum of Total for LSOAs in Harrow    #equivalent using dot notation is harrow_sumTot = harrow.Total.sum() 

#convert to float when making calculation
harrow_propG2 =  float(harrow_sumG2) / float(harrow_sumTot)          #calculate proportion 
print "The proportion of Harrow in Group2 is:", str(harrow_propG2)   #print nicely

Run the code above and check that you find that the proportion is 0.236. Also check you understand how (and why) harrow_SumG2 and harrow_sumTot were created – they were created by indexing the bSums DataFrame created by aggregate. What happens if you do not force these to be float when calculating the proportion?

# Summary 

In this practical we have....

You now have the 'full' data set that you can use for your final report. So start exploring!

If you want to join your own data for analysis in your final report, please discuss with James or Jon before doing so. 